### PR TITLE
Bugfix/b(#2778) tooltip position when container uses transform scale

### PR DESCRIPTION
**Pull request by MateuszTrN**
Fixes detection of a tooltip mouse event

## Description

When a parent container used transform: scale(Number) Css - the tooltip mouse event was fired up in the incorrect mouse position 

## Related Issue

https://github.com/recharts/recharts/issues/2778

## Motivation and Context

This changes allows user to style parent container with transform style css property without affecting tooltip behaviour

## How Has This Been Tested?

npm run test, npm run storybook and did some manual testing. I also created a dedicated story.
The changes should not affect other parts of the code - it was a change within 1 function, that already had tests around it. It introduces new parameter to a "inRange" function that have default value if wasn't passed. The API of that function was not changed, only extended. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have added a story to the storybook
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ckifer** [comment]
Thanks for the contribution! Do you mind resolving the conflict?

File: src/chart/generateCategoricalChart.tsx [hunk 1]
```diff
@@ -1253,7 +1253,13 @@ export const generateCategoricalChart = ({
 
       const containerOffset = getOffset(this.container);
       const e = calculateChartCoordinate(event, containerOffset);
-      const rangeObj = this.inRange(e.chartX, e.chartY);
+
+      const element = this.container;
+      const boundingRectWidth = element?.getBoundingClientRect()?.width;
+      const { offsetWidth } = element;
+      const scale = boundingRectWidth / offsetWidth || 1;
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1260]: Just checking, this division is safe if it results in `NaN` correct?

File: src/chart/generateCategoricalChart.tsx [hunk 2]
```diff
@@ -1343,23 +1349,28 @@ export const generateCategoricalChart = ({
       ];
     }
 
-    inRange(x: number, y: number): any {
+    inRange(x: number, y: number, scale = 1): any {
       const { layout } = this.props;
 
+      const [scaledX, scaledY] = [x / scale, y / scale];
```
Comment (ckifer) [inline src/chart/generateCategoricalChart.tsx:1355]: what happens if these result in NaN?

**Review by ckifer — COMMENTED** [review]
Nice! Just a few questions

File: src/chart/generateCategoricalChart.tsx (hunk 1 above)
Comment (MateuszTrN) [reply src/chart/generateCategoricalChart.tsx:1260]: yes, line 1260 returns 1 if the div is NaN (or 0 - which I also assumed is a value that does not make sense)

File: src/chart/generateCategoricalChart.tsx (hunk 2 above)
Comment (MateuszTrN) [reply src/chart/generateCategoricalChart.tsx:1355]: I can make another check here too, but I assumed I dealt with that in the parent scope, before passing the arg here. as this is internal function that is not exposed so I did not do a check again, but I can add it if you think it's better that way

Comment (ckifer) [reply src/chart/generateCategoricalChart.tsx:1355]: Nah it should be fine. I do feel though like this function could just be totally taken out of this file and probably have some different, more targeted tests added to it. Probably an improvement for another time though

**Review by ckifer — APPROVED** [review]
LGTM, will wait for @nikolasrieble or someone else to take a look as well

**Review by nikolasrieble — APPROVED** [review]
A great contribution that solves a long standing problem! Thank you a lot. 
I tested it and could not find any edge cases where this would not work. 

The only constructive feedback I have here is: 
I wish we could extract useful helper function instead of extending the complexity on the existing level of abstraction in `src/chart/generateCategoricalChart.tsx`. Ideally every PR from now on would only take code out of generateCategoricalChart (and make it nicely unit testable). 

That being said, this is great to go, and we should focus on making steps forward. This certainly is a step forward, thank you ❤️

**Pull request merged**